diff --git a/Makefile b/Makefile
index 6b4eb85..7281477 100644
--- a/Makefile
+++ b/Makefile
@@ -103,6 +103,20 @@ else ifeq ($(platform), miyoo)
    CFLAGS += -march=armv5te -mtune=arm926ej-s
    CFLAGS += -fomit-frame-pointer -ffast-math  
    
+# trimui
+else ifeq ($(platform), trimui)
+   TARGET := $(TARGET_NAME)_libretro.so
+   CC = $(CROSS_COMPILE)gcc
+   CXX = $(CROSS_COMPILE)g++
+   AR = $(CROSS_COMPILE)ar
+   SHARED := -shared -Wl,-version-script=link.T -Wl,-no-undefined
+   LDFLAGS += -fno-PIC -flto
+   CFLAGS += -fomit-frame-pointer -ffast-math -mcpu=arm926ej-s -mtune=arm926ej-s -fno-PIC -flto
+   OPTIMIZE := -Ofast -DNDEBUG
+   ifeq (,$(DEBUG))
+      LDFLAGS += -s
+   endif
+
 # iOS
 else ifneq (,$(findstring ios,$(platform)))
 
@@ -532,6 +546,8 @@ endif
 
 ifeq ($(DEBUG), 1)
 	CFLAGS += -O0 -g
+else ifneq (,$(OPTIMIZE))
+	CFLAGS += $(OPTIMIZE)
 else
 	CFLAGS += -O2 -DNDEBUG
 endif
@@ -546,6 +562,13 @@ else
 endif
 endif
 
+ifeq ($(PROFILE), GENERATE)
+	CFLAGS	+= -fprofile-generate=./profile/$(TARGET_NAME)
+	LDFLAGS	+= -lgcov
+else ifeq ($(PROFILE), APPLY)
+	CFLAGS	+= -fprofile-use -fprofile-dir=../profile/$(TARGET_NAME) -fbranch-probabilities
+endif
+
 DEFINES += -D__LIBRETRO__ $(PLATFORM_DEFINES)
 
 CFLAGS   += $(fpic) $(DEFINES)
