diff --git a/Makefile b/Makefile
index fa45fb3..9e20a41 100644
--- a/Makefile
+++ b/Makefile
@@ -39,6 +39,7 @@ ifeq ($(platform), unix)
 	ifneq ($(findstring Haiku,$(shell uname -a)),)
 		LIBM :=
 	endif
+	CFLAGS += -D_GNU_SOURCE
 
 # OS X
 else ifeq ($(platform), osx)
@@ -214,6 +215,19 @@ else ifeq ($(platform), switch)
         include $(LIBTRANSISTOR_HOME)/libtransistor.mk
         STATIC_LINKING=1
 
+else ifeq ($(platform), trimui)
+	TARGET := $(TARGET_NAME)_libretro.so
+	CC = $(CROSS_COMPILE)gcc
+	CXX = $(CROSS_COMPILE)g++
+	AR = $(CROSS_COMPILE)ar
+	SHARED := -shared -Wl,--version-script=link.T
+	CFLAGS += -fomit-frame-pointer -march=armv5te -mtune=arm926ej-s -ffast-math -fno-PIC -flto -D_GNU_SOURCE
+	CXXFLAGS += -fomit-frame-pointer -march=armv5te -mtune=arm926ej-s -ffast-math -fno-PIC -flto -D_GNU_SOURCE
+	LDFLAGS += -flto
+	ifeq (,$(DEBUG))
+		LDFLAGS += -s
+	endif
+
 # ARM
 else ifneq (,$(findstring armv,$(platform)))
 	TARGET := $(TARGET_NAME)_libretro.so
