diff --git a/Makefile.libretro b/Makefile.libretro
index 919e0c1..e3bcd36 100644
--- a/Makefile.libretro
+++ b/Makefile.libretro
@@ -517,6 +517,23 @@ else ifeq ($(platform), miyoo)
 	LOW_MEMORY = 1
 	NO_ARM_ASM = 1
 
+# TRIMUI
+else ifeq ($(platform), trimui)
+	TARGET := $(TARGET_NAME)_libretro.so
+	CC = $(CROSS_COMPILE)gcc
+	CXX = $(CROSS_COMPILE)g++
+	AR = $(CROSS_COMPILE)ar
+	SHARED := -shared -Wl,--no-undefined
+  DONT_COMPILE_IN_ZLIB = 1
+	CFLAGS += -fomit-frame-pointer -ffast-math -march=armv5te -mtune=arm926ej-s -D__GCW0__
+	CFLAGS += -fno-PIC -flto
+	LDFLAGS += -flto
+	HAVE_ARMv6 = 0
+	LOW_MEMORY = 1
+	ifeq (,$(DEBUG))
+		LDFLAGS += -s
+	endif
+
 # Windows MSVC 2017 all architectures
 else ifneq (,$(findstring windows_msvc2017,$(platform)))
 	NO_GCC := 1
@@ -720,6 +737,13 @@ else
 	endif
 endif
 
+ifeq ($(PROFILE), GENERATE)
+	CFLAGS	+= -fprofile-generate=./profile/$(TARGET_NAME)
+	LDFLAGS	+= -lgcov
+else ifeq ($(PROFILE), APPLY)
+	CFLAGS	+= -fprofile-use -fprofile-dir=../profile/$(TARGET_NAME) -fbranch-probabilities
+endif
+
 ifneq (,$(findstring msvc,$(platform)))
 	OBJOUT = -Fo
 	LINKOUT = -out:
