diff --git a/libretro_core_options.h b/libretro_core_options.h
index 41d5ff1..f41ee18 100644
--- a/libretro_core_options.h
+++ b/libretro_core_options.h
@@ -68,7 +68,7 @@ struct retro_core_option_v2_definition option_defs_us[] = {
       "snes9x_2005_region",
       "Console Region",
       NULL,
-      "Specify which region the system is from. 'PAL' is 50hz, 'NTSC' is 60hz. Games will run faster or slower than normal if the incorrect region is selected.",
+      "'PAL' is 50hz, 'NTSC' is 60hz. Games will run faster or slower than normal if the incorrect region is selected.",
       NULL,
       NULL,
       {
@@ -83,7 +83,7 @@ struct retro_core_option_v2_definition option_defs_us[] = {
       "snes9x_2005_frameskip",
       "Frameskip",
       NULL,
-      "Skip frames to avoid audio buffer under-run (crackling). Improves performance at the expense of visual smoothness. 'Auto' skips frames when advised by the frontend. 'Manual' utilises the 'Frameskip Threshold (%)' setting.",
+      "Skip frames to avoid audio crackling. Improves performance at the expense of visual smoothness.",
       NULL,
       NULL,
       {
@@ -92,13 +92,13 @@ struct retro_core_option_v2_definition option_defs_us[] = {
          { "auto_threshold", "Threshold" },
          { NULL, NULL },
       },
-      "disabled"
+      "auto"
    },
    {
       "snes9x_2005_frameskip_threshold",
-      "Frameskip Threshold (%)",
+      "FS Threshold (%)",
       NULL,
-      "When 'Frameskip' is set to 'Manual', specifies the audio buffer occupancy threshold (percentage) below which frames will be skipped. Higher values reduce the risk of crackling by causing frames to be dropped more frequently.",
+      "When 'Frameskip' is set to 'Threshold', sets how low the audio buffer can get before frames will be skipped.",
       NULL,
       NULL,
       {
@@ -124,7 +124,7 @@ struct retro_core_option_v2_definition option_defs_us[] = {
    },
    {
       "snes9x_2005_frameskip_interval",
-      "Frameskip Interval",
+      "FS Interval",
       NULL,
       "The maximum number of frames that can be skipped before a new frame is rendered.",
       NULL,
@@ -147,9 +147,9 @@ struct retro_core_option_v2_definition option_defs_us[] = {
    },
    {
       "snes9x_2005_overclock_cycles",
-      "Reduce Slowdown (Hack, Unsafe, Restart)",
+      "Overclock (Restart)",
       NULL,
-      "Many games for the SNES suffered from slowdown due to the weak main CPU. This option helps allievate that at the cost of possible bugs. COMPATIBLE: Reduce slowdown but keep as much game compatibility as much as possible. MAX: Reduce slowdown as much as possible but will break more games.",
+      "Alleviate normal SNES slowdown. Compatible keeps as much compatibility as possible. Max will reduce more slowdown but break more games.",
       NULL,
       "hacks",
       {
@@ -162,9 +162,9 @@ struct retro_core_option_v2_definition option_defs_us[] = {
    },
    {
       "snes9x_2005_reduce_sprite_flicker",
-      "Reduce Flickering (Hack, Unsafe)",
+      "Reduce Flicker",
       NULL,
-      "Increases number of sprites that can be drawn simultaneously on screen. Reduces flickering in games.",
+      "Raises sprite limit to reduce flickering in games.",
       NULL,
       "hacks",
       {
