diff --git a/Makefile b/Makefile
index a0ed865..95c1438 100644
--- a/Makefile
+++ b/Makefile
@@ -470,6 +470,20 @@ else ifeq ($(platform), miyoo)
 	HAVE_DYNAREC := 1
 	CPU_ARCH := arm
 
+else ifeq ($(platform), trimui)
+  TARGET := $(TARGET_NAME)_libretro.so
+	CC = $(CROSS_COMPILE)gcc
+  SHARED := -shared -Wl,--version-script=link.T -mcpu=arm926ej-s -mtune=arm926ej-s
+	CFLAGS += -ffast-math -fdata-sections -ffunction-sections -flto -fno-PIC -DOVERCLOCK_60FPS
+	LDFLAGS += -flto
+	ifeq (,$(DEBUG))
+		LDFLAGS += -s
+	endif
+	HAVE_DYNAREC := 1
+	CPU_ARCH_ARM_BLENDING_OPTS := 1
+	CPU_ARCH := arm
+	OPTIMIZE := -Ofast -DNDEBUG
+
 # Windows
 else
 	TARGET := $(TARGET_NAME)_libretro.dll
@@ -491,7 +505,14 @@ endif
 ifeq ($(DEBUG), 1)
 	OPTIMIZE      := -O0 -g
 else
-	OPTIMIZE      := -O3 -DNDEBUG
+	OPTIMIZE      ?= -O3 -DNDEBUG
+endif
+
+ifeq ($(PROFILE), GENERATE)
+	CFLAGS	+= -fprofile-generate=./profile/gpsp
+	LDFLAGS	+= -lgcov
+else ifeq ($(PROFILE), APPLY)
+	CFLAGS	+= -fprofile-use -fprofile-dir=../profile/gpsp -fbranch-probabilities
 endif
 
 DEFINES := -DHAVE_STRINGS_H -DHAVE_STDINT_H -DHAVE_INTTYPES_H -D__LIBRETRO__ -DINLINE=inline -Wall
