diff --git a/libretro/libretro.c b/libretro/libretro.c
index 59c1b33..be1709c 100644
--- a/libretro/libretro.c
+++ b/libretro/libretro.c
@@ -1106,7 +1106,7 @@ void retro_run(void)
                         1 : 0;
 
             if (!skip_next_frame ||
-                (frameskip_counter >= FRAMESKIP_MAX))
+                (frameskip_counter >= frameskip_interval))
             {
                skip_next_frame   = 0;
                frameskip_counter = 0;
@@ -1123,7 +1123,7 @@ void retro_run(void)
                               1 : 0;
 
             if (!skip_next_frame ||
-                (frameskip_counter >= FRAMESKIP_MAX))
+                (frameskip_counter >= frameskip_interval))
             {
                skip_next_frame   = 0;
                frameskip_counter = 0;
