diff --git a/Makefile.libretro b/Makefile.libretro
index e4575bc..7f96e07 100644
--- a/Makefile.libretro
+++ b/Makefile.libretro
@@ -327,7 +327,21 @@ else ifeq ($(platform), miyoo)
    SHARED := -shared -Wl,--no-undefined -Wl,--version-script=link.T
    CFLAGS += -ffast-math -march=armv5te -mtune=arm926ej-s -fomit-frame-pointer
    ENDIANNESS_DEFINES += -DLSB_FIRST -DBYTE_ORDER=LITTLE_ENDIAN -DALIGN_LONG
-   
+
+else ifeq ($(platform), trimui)
+	TARGET := $(TARGET_NAME)_libretro.so
+	CC = $(CROSS_COMPILE)gcc
+	CXX = $(CROSS_COMPILE)g++
+	AR = $(CROSS_COMPILE)ar
+	SHARED := -shared -Wl,--no-undefined -Wl,--version-script=link.T
+	CFLAGS += -fomit-frame-pointer -march=armv5te -mtune=arm926ej-s -ffast-math -fno-PIC -flto
+	CXXFLAGS += -fomit-frame-pointer -march=armv5te -mtune=arm926ej-s -ffast-math -fno-PIC -flto
+	LDFLAGS += -flto
+	ifeq (,$(DEBUG))
+		LDFLAGS += -s
+	endif
+	ENDIANNESS_DEFINES += -DLSB_FIRST -DBYTE_ORDER=LITTLE_ENDIAN -DALIGN_LONG
+
 # else ifneq (,$(findstring armv,$(platform)))
 #    TARGET := $(TARGET_NAME)_libretro.so
 #    fpic := -fPIC
