diff --git a/Makefile.libretro b/Makefile.libretro
index f41fcc2..50a1134 100644
--- a/Makefile.libretro
+++ b/Makefile.libretro
@@ -368,6 +368,21 @@ else ifeq ($(platform), gcw0)
    CFLAGS += -fomit-frame-pointer -ffast-math -march=mips32 -mtune=mips32r2 -mhard-float
    CXXFLAGS += $(CFLAGS)
 
+# Trimui
+else ifeq ($(platform), trimui)
+	TARGET := $(TARGET_NAME)_libretro.so
+	CC = $(CROSS_COMPILE)gcc
+	CXX = $(CROSS_COMPILE)g++
+	AR = $(CROSS_COMPILE)ar
+	SHARED := -shared -Wl,--version-script=$(version_script)
+	LDFLAGS += -fno-PIC -flto
+	PLATFORM_DEFINES := -DCC_RESAMPLER -DCC_RESAMPLER_NO_HIGHPASS
+	CFLAGS += -fomit-frame-pointer -ffast-math -mcpu=arm926ej-s -mtune=arm926ej-s -fno-PIC -flto
+	OPTIMIZE := -Ofast -DNDEBUG
+	ifeq (,$(DEBUG))
+		LDFLAGS += -s
+	endif
+
 # RETROFW
 else ifeq ($(platform), retrofw)
    TARGET := $(TARGET_NAME)_libretro.so
@@ -622,10 +637,23 @@ ifneq (,$(findstring msvc,$(platform)))
    CFLAGS   += -MT
    CXXFLAGS += -MT
 endif
+ifneq (,$(OPTIMIZE))
+	CFLAGS += $(OPTIMIZE)
+	CXXFLAGS += $(OPTIMIZE)
+else
    CFLAGS   += -O2 -DNDEBUG
    CXXFLAGS += -O2 -DNDEBUG
 endif
 
+endif
+
+ifeq ($(PROFILE), GENERATE)
+	CFLAGS	+= -fprofile-generate=./profile/gambatte
+	LDFLAGS	+= -lgcov
+else ifeq ($(PROFILE), APPLY)
+	CFLAGS	+= -fprofile-use -fprofile-dir=../profile/gambatte -fbranch-probabilities
+endif
+
 ifeq (,$(findstring msvc,$(platform)))
 CXXFLAGS += -fno-exceptions -fno-rtti
 CXXFLAGS += -std=c++98
