diff --git a/libretro/libretro.c b/libretro/libretro.c
index 5f947ef..dc48f3d 100644
--- a/libretro/libretro.c
+++ b/libretro/libretro.c
@@ -105,10 +105,6 @@ typedef enum
    FRAMESKIP_FIXED_INTERVAL
 } frameskip_type_t;
 
-/* Maximum number of consecutive frames that
- * can be skipped */
-#define FRAMESKIP_MAX 30
-
 static frameskip_type_t frameskip_type           = FRAMESKIP_NONE;
 static unsigned frameskip_threshold              = 0;
 static unsigned frameskip_interval               = 0;
@@ -638,7 +634,7 @@ void retro_run (void)
                   retro_audio_buff_underrun;
 
             if (!skip_frame ||
-                (frameskip_counter >= FRAMESKIP_MAX))
+                (frameskip_counter >= frameskip_interval))
             {
                skip_frame        = false;
                frameskip_counter = 0;
@@ -652,7 +648,7 @@ void retro_run (void)
                   (retro_audio_buff_occupancy < frameskip_threshold);
 
             if (!skip_frame ||
-                (frameskip_counter >= FRAMESKIP_MAX))
+                (frameskip_counter >= frameskip_interval))
             {
                skip_frame        = false;
                frameskip_counter = 0;
