diff --git a/libretro/libretro.c b/libretro/libretro.c
index 81ce4aa..bc64376 100644
--- a/libretro/libretro.c
+++ b/libretro/libretro.c
@@ -339,7 +339,7 @@ void retro_init (void)
 {
    static const struct retro_variable vars[] =
    {
-      { "snes9x2002_overclock_cycles", "Reduce Slowdown (Hack, Unsafe, Restart); disabled|compatible|max" },
+      { "snes9x2002_overclock_cycles", "Overclock (Restart); disabled|compatible|max" },
       { NULL, NULL },
    };
 
